File: packages/agents/src/opencode.ts
/**
 * OpenCode adapter.
 *
 * Path and hook choices are isolated to this module: when OpenCode's real
 * on-disk layout is confirmed, only the values below change, never their
 * consumers.
 *
 * - Skills: project `<project>/.opencode/skills/<name>/`, global
 *   `~/.config/opencode/skills/<name>/` (OpenCode follows the XDG convention for
 *   its per-user configuration).
 * - Hooks: `delimited-text` into the OpenCode config file. OpenCode's config is
 *   a comment-capable format, so SkillKeeper inserts an owned, comment-delimited
 *   region (comment token `#`) rather than merging JSON. This also exercises the
 *   delimited-text strategy across the built-in adapter set.
 */

import type { AgentTarget, HostEnv } from '@skillkeeper/core';
import { makeAdapter } from './makeAdapter.js';
import { baseDir, joinPath, requireProjectDir } from './paths.js';

/** OpenCode's base config directory differs between project and global scope. */
function opencodeDir(target: AgentTarget, env: HostEnv): string {
  return target.scope === 'project'
    ? joinPath(requireProjectDir(env), '.opencode')
    : joinPath(env.homeDir, '.config', 'opencode');
}

export const opencodeAdapter = makeAdapter({
  kind: 'opencode',
  skillsRoot: (target, env) => joinPath(opencodeDir(target, env), 'skills'),
  availabilityDir: (env) => joinPath(env.homeDir, '.config', 'opencode'),
  guidanceFile: async (target, env) => joinPath(baseDir(target, env), 'AGENTS.md'),
  hook: {
    strategy: 'delimited-text',
    commentToken: '#',
    async resolveTargetFile(target: AgentTarget, env: HostEnv): Promise<string> {
      return joinPath(opencodeDir(target, env), 'opencode.json');
    },
  },
});

File: crates/skillkeeper-agents/src/opencode.rs
//! OpenCode adapter. Ported from `packages/agents/src/opencode.ts`.
//!
//! - Skills: project `<project>/.opencode/skills/<name>/`, global
//!   `~/.config/opencode/skills/<name>/`.
//! - Hooks: `delimited-text` into `opencode.json`. OpenCode's config is a
//!   comment-capable format, so an owned, comment-delimited region (comment
//!   token `#`) is inserted rather than merging JSON.

use skillkeeper_core::ports::{HostEnv, PortResult};

use crate::adapter::{make_adapter, AdapterSpec, AgentAdapter, HookCapability};
use crate::model::{AgentKind, AgentTarget, HookStrategy, Scope};
use crate::paths::{base_dir, join_path, require_project_dir};

/// OpenCode's base config directory differs between project and global scope.
fn opencode_dir(target: &AgentTarget, env: &dyn HostEnv) -> PortResult<String> {
    match target.scope {
        Scope::Project => Ok(join_path(&[&require_project_dir(env)?, ".opencode"])),
        Scope::Global => Ok(join_path(&[env.home_dir(), ".config", "opencode"])),
    }
}

/// Build the OpenCode adapter.
pub fn opencode_adapter() -> AgentAdapter {
    make_adapter(AdapterSpec {
        kind: AgentKind::Opencode,
        skills_root: Box::new(|target, env| {
            Ok(join_path(&[&opencode_dir(target, env)?, "skills"]))
        }),
        availability_dir: Box::new(|env| join_path(&[env.home_dir(), ".config", "opencode"])),
        discovery_group: None,
        guidance_file: Box::new(|_fs, target, env| {
            Ok(join_path(&[&base_dir(target, env)?, "AGENTS.md"]))
        }),
        hook: HookCapability::new(
            HookStrategy::DelimitedText,
            Some("#".to_string()),
            None,
            Box::new(|target, env| Ok(join_path(&[&opencode_dir(target, env)?, "opencode.json"]))),
        ),
    })
}
